cardano-testnet/test/Test/Cli/QuerySlotNumber.hs

```diff
@@ -0,0 +1,156 @@
+
+  -- there's an offset for start time, so we have to account for that here
+  let startTime = ((-1) * startTimeOffsetSeconds) `DT.addUTCTime` sgSystemStart
+      slotLength = fromNominalDiffTimeMicro sgSlotLength
+      -- how many slots can the checked value differ from
+      -- we have 1s precision for UTC timestamp CLI argument, so this value tells how many slots in 1s can be
+      slotPrecision = round $ 1 / slotLength
+      -- first era has epochsize 100 - that's a different value than in ShelleyGenesis
+      epochSize = 100
+      passedSlots = 100 :: Int
+      passedTime = fromIntegral passedSlots * fromNominalDiffTimeMicro sgSlotLength
+
+  poolNode1 <- H.headM poolNodes
+  poolSprocket1 <- H.noteShow $ nodeSprocket $ poolRuntime poolNode1
+  env <- H.evalIO getEnvironment
+  execConfig <- H.noteShow H.ExecConfig
+    { H.execConfigEnv = Last $ Just $
+      [ ("CARDANO_NODE_SOCKET_PATH", IO.sprocketArgumentName poolSprocket1)
+      ]
+      -- The environment must be passed onto child process on Windows in order to
+      -- successfully start that process.
+      <> env
+    , H.execConfigCwd = Last $ Just tempBaseAbsPath
+    }
+
+  -- {{{ Test cases
+  id do
+    H.note_ "Double check that start time offset is larger than the passed time we test"
+    H.diff startTimeOffsetSeconds (>) passedTime
+
+  id do
+    H.note_ "Try to retrieve slot 5s before genesis"
+    testTime <- H.note . formatTime $ (-5) `DT.addUTCTime` startTime
+    (result, _) <- H.runTestT $ execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    H.assert $ isLeft result
+
+  id do
+    H.note_ "Retrieve slot number for the start time"
+    testTime <- H.note $ formatTime startTime
+    let expectedSlot = 0
+    slot <- readM @Int =<< execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    isWithinDisk slot expectedSlot slotPrecision
+
+  id do
+    H.note_ "Retrieve slot number for some delay"
+    testTime <- H.note . formatTime $ passedTime `DT.addUTCTime` startTime
+    let expectedSlot = passedSlots
+    slot <- readM @Int =<< execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    isWithinDisk slot expectedSlot slotPrecision
+
+  id do
+    H.note_ "Retrieve slot number for next epoch"
+    let epochDuration = slotLength * fromIntegral epochSize
+    testTime <- H.note . formatTime $ epochDuration `DT.addUTCTime` startTime
+    let expectedSlot = epochSize
+    slot <- readM @Int =<< execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    isWithinDisk slot expectedSlot slotPrecision
+
+  id do
+    H.note_ "Try to retrieve slot far in the future"
+    -- 10 epochs ahead
+    let timeOffset = slotLength * fromIntegral epochSize * 10
```

Comment:
The first era has `epochSize = 100` . But it appears that I am able to query and get results for slots up to around 600th slot. The system is at ~150th slot during this invocation, so it's ~450 slots ahead of time. That's definitely further than `N+1` epoch, and `N+2`  epoch for size of 100. On the other side it's less than the end of `N+1` epoch for `epochSize = 600` - which is supposed to be set by test options for Babbage.  So I've just picked a very large value to make sure that we're getting exception when looking past horizon. But I don't know how to calculate how far the horizon is on the testnet.

This works correctly on mainnet though.

Reply:
I think this should be a comment in the test.